### DISCUSSION TITLE
Underscore is not recognized as substitution separator

### DISCUSSION THREAD

**Discussion by Vvalter** [JetBrains/ideavim#393] [Q&A] [answered]
According to `:help E146` it's possible in vim to use other separators with substitution than `/`. In particular, it's possible in vim to use underscore as separator. E.g. `:%s_a_b_g` will replace all occurrences of a with b in the whole document.
Ideavim does not accept underscore as separator (other separators work fine).

The reason is that `s_a_b_g` gets lexed as `IDENTIFIER_ANY_CASE_WITH_DIGITS_AND_UNDERSCORES` which matches [`OtherCommand`](https://github.com/JetBrains/ideavim/blob/36002def70a63a8ac0443e8babf2ca4a32ba9f9e/src/main/antlr/Vimscript.g4#L136) instead of `CommandWithBars`.
I'm not that familiar with ANTLR, but I don't see a clean way to fix this in the grammar. Manually reparsing the token later only seems like a too ugly hack to me. Does anybody see a clean way to fix this?

By the way, vim explicitly forbids `\`, `"`, and `|` as separators, but Ideavim accepts them and does either no replacements or weird replacements. E.g. `:%s"a"b"g` seems to insert underscores in front of single quotes?

**Comment by AlexPl292** [comment]
@lippfi could you please take a look on these two issues?

**Comment by lippfi** [comment, marked as answer]
The fix is available in the IdeaVim 1.8.0. Thank your for your report!

> By the way, vim explicitly forbids \, ", and | as separators, but Ideavim accepts them and does either no replacements or weird replacements. E.g. :%s"a"b"g seems to insert underscores in front of single quotes?

In such cases IdeaVim (and Vim) uses the last used pattern

**Reply by Vvalter** [reply]
Thanks for the quick fix :1st_place_medal:
